cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/partitioned/PartitionConsumer.java

```diff
@@ -55,4 +56,13 @@
    * @param succeeded whether or not processing of the specified partitions was successful
    */
   void onFinish(List<? extends Partition> partitions, boolean succeeded);
+
+  /**
+   * Same as {@link #onFinish(List, boolean)}, but allows specifying {@link PartitionKey}s
+   * instead of {@link co.cask.cdap.api.dataset.lib.Partition}s.
+   *
+   * @param partitionKeys list of partition keys to mark as either succeeded or failed processing
+   * @param succeeded whether or not processing of the specified partitions was successful
+   */
+  void onFinishWithKeys(List<? extends PartitionKey> partitionKeys, boolean succeeded);
```

Comment:
Just tried it and making onFinishWithKeys take Iterable instead of List won't work, because we define a doFinish method in AbstractPartitionConsumer which gets called from the `onFinish` and `onFinishWithKeys`.

Its signature:
`public abstract void doFinish(ConsumerWorkingSet workingSet, List<? extends PartitionKey> partitionKeys, boolean succeeded);`